amethyst_ui/src/drag.rs

```diff
@@ -0,0 +1,123 @@
+use serde::{Deserialize, Serialize};
+use std::marker::PhantomData;
+
+use amethyst_core::{
+    ecs::{
+        Component, DenseVecStorage, Entities, Entity, Join, Read, ReaderId, System, SystemData,
+        Write, WriteStorage,
+    },
+    math::Vector2,
+    shrev::EventChannel,
+};
+use amethyst_derive::SystemDesc;
+use amethyst_input::{BindingTypes, InputEvent};
+
+use crate::{UiEvent, UiEventType, UiTransform};
+
+/// Component that denotes whether a given ui widget is draggable.
+/// Requires UiTransform to work, and its expected way of usage is
+/// through UiTransformData prefab.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Draggable {
+    /// When true, the widget is being dragged. When false, it's not.
+    pub being_dragged: bool,
+}
+
+impl Component for Draggable {
+    type Storage = DenseVecStorage<Self>;
+}
+
+#[derive(Debug, SystemDesc)]
+#[system_desc(name(DragWidgetSystemDesc))]
+pub struct DragWidgetSystem<T: BindingTypes> {
+    #[system_desc(event_channel_reader)]
+    input_reader_id: ReaderId<InputEvent<T>>,
+
+    #[system_desc(event_channel_reader)]
+    ui_reader_id: ReaderId<UiEvent>,
+
+    phantom: PhantomData<T>,
+}
+
+impl<T> DragWidgetSystem<T>
+where
+    T: BindingTypes,
+{
+    pub fn new(input_reader_id: ReaderId<InputEvent<T>>, ui_reader_id: ReaderId<UiEvent>) -> Self {
+        Self {
+            input_reader_id,
+            ui_reader_id,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'s, T> System<'s> for DragWidgetSystem<T>
+where
+    T: BindingTypes,
+{
+    type SystemData = (
+        Entities<'s>,
+        Read<'s, EventChannel<InputEvent<T>>>,
+        Write<'s, EventChannel<UiEvent>>,
+        WriteStorage<'s, Draggable>,
+        WriteStorage<'s, UiTransform>,
+    );
+
+    fn run(
+        &mut self,
+        (entities, input_events, mut ui_events, mut draggables, mut ui_transforms): Self::SystemData,
+    ) {
+        let mut click_stopped: Vec<Entity> = Vec::new();
+
+        for event in ui_events.read(&mut self.ui_reader_id) {
+            match event.event_type {
+                UiEventType::ClickStart => {
+                    if let Some(draggable) = draggables.get_mut(event.target) {
+                        draggable.being_dragged = true;
+                    }
+                }
+                UiEventType::ClickStop => {
+                    if let Some(draggable) = draggables.get(event.target) {
+                        if draggable.being_dragged {
+                            click_stopped.push(event.target);
```

Comment:
event.target contains the entity that we stopped dragging, not the entity that we dropped onto.
I think that its this entity (the dragged one) that you are inserting into the dropped_on field here https://github.com/amethyst/amethyst/pull/2080/files#diff-896eb5ae9178dff5949020bc851a076dR117 instead of the actual target we are dropping on.

Reply:
Interesting. I did not know that dropped_on field was supposed to be like that. But that does raise some issues. What if the draggable entity is stopped being dragged, but there's no entity on which it was dropped onto? What if there are multiple entities a given draggable entity was dropped onto? I'm not entirely sure how I should handle such scenarios.

Reply:
As far as I'm concerned when one stops dragging the entity and it had not entered any drop zone it should be returned to it's initial position (the position before the dragging started). Basically this means that we need to save the initial position when the drag starts.

Can't really imagine what do you mean by the latter scenario ? I would say you would drop it to the closest drop zone ?